core/src/verified_vote_packets.rs

```diff
@@ -1,930 +0,0 @@
-                        // it has a later timestamp (refreshed vote)
-                        //
-                        // Timestamp can be None if something was wrong with the senders clock.
-                        // We directly compare as Options to ensure that votes with proper
-                        // timestamps have precedence (Some is > None).
-                        if slot > latest_gossip_slot
-                            || ((slot == latest_gossip_slot) && (timestamp > latest_timestamp))
-                        {
-                            self.0.insert(
-                                vote_account_key,
-                                FullTowerVote(GossipVote {
-                                    slot,
-                                    hash,
-                                    packet_batch,
-                                    signature,
-                                    timestamp,
-                                }),
-                            );
-                        }
-                    }
-                    _ => {
-                        if let Some(FullTowerVote(gossip_vote)) = self.0.get_mut(&vote_account_key)
-                        {
-                            if slot > gossip_vote.slot {
-                                warn!(
-                                        "Originally {} submitted full tower votes, but now has reverted to incremental votes. Converting back to old format.",
-                                        vote_account_key
-                                    );
-                                let mut votes = BTreeMap::new();
-                                let GossipVote {
-                                    slot,
-                                    hash,
-                                    packet_batch,
-                                    signature,
-                                    ..
-                                } = std::mem::take(gossip_vote);
-                                votes.insert((slot, hash), (packet_batch, signature));
-                                self.0.insert(vote_account_key, IncrementalVotes(votes));
-                            } else {
-                                continue;
-                            }
-                        };
-                        let validator_votes: &mut BTreeMap<(Slot, Hash), (PacketBatch, Signature)> =
-                            match self
-                                .0
-                                .entry(vote_account_key)
-                                .or_insert(IncrementalVotes(BTreeMap::new()))
-                            {
-                                IncrementalVotes(votes) => votes,
-                                FullTowerVote(_) => continue, // Should never happen
-                            };
-                        validator_votes.insert((slot, hash), (packet_batch, signature));
-                        if validator_votes.len() > MAX_VOTES_PER_VALIDATOR {
```

Comment:
So I'm a bit curious about the change as it relates to this line.

With this code we used to store up to 1000 votes per validator.
Votes on our fork would get sent to banking stage and the others remain here in case we switch forks.
But now, we bypass this and all just go directly to banking stage.
Since banking stage only stores the **most recent** vote (to my knowledge) we no longer have this reserve for "in case we switch forks".

Reply:
This code was to deal with pre `VoteStateUpdate`, back when incremental votes were being sent. This business with the `IncrementalVotes` and `FullTowerVote` was to deal with the feature activation and gracefully handle a possible rollback if the feature was busted.

Previously when votes where incremental we stored up to 1000 of the last votes, as we wanted to land as many as possible to ensure that the on chain tower matched the local tower. Now we only store one vote. Even in the situation of switching forks, the most recent vote suffices.  

Non modded validators do not send incremental votes, so I think it's fine to just store the most recent vote in banking stage. There is an effort I am working on to remove incremental votes from the wire all together, but for now treating them incorrectly is fine.